src/gui/properties/piecesmapwidget.cpp

```diff
@@ -0,0 +1,195 @@
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "piecesmapwidget.h"
+
+#include <QtMath>
+#include <QPainter>
+#include <QResizeEvent>
+#include <QScrollBar>
+#include <QProgressBar>
+#include <QProxyStyle>
+#include <QStyleOptionProgressBar>
+
+#include "base/bittorrent/torrent.h"
+#include "propertieswidget.h"
+#include "base/bittorrent/session.h"
+#include "base/bittorrent/infohash.h"
+
+
+PiecesMapWidget::PiecesMapWidget(QScrollArea *parent)
+    : QWidget(parent)
+    , m_scrollArea(parent)
+{
+    m_piecesCount = 0;
+    m_piecePixelSize = 15;
+    m_spacePixelSize = 5;
+    m_height = 0;
+
+    BitTorrent::Session* session = BitTorrent::Session::instance();
+    QObject::connect(session, &BitTorrent::Session::pieceFinished, this, &PiecesMapWidget::handlePieceFinished);
+
+    QProgressBar dummyProgressBar;
+    dummyProgressBar.setFixedSize(m_piecePixelSize, m_piecePixelSize);
+    dummyProgressBar.setTextVisible(false);
+    dummyProgressBar.setMinimum(0);
+    dummyProgressBar.setMaximum(100);
+#if (defined(Q_OS_WIN) || defined(Q_OS_MACOS))
+    auto *fusionStyle = new QProxyStyle {"fusion"};
+    fusionStyle->setParent(&dummyProgressBar);
+    dummyProgressBar.setStyle(fusionStyle);
+#endif
+    QPixmap pixmap(m_piecePixelSize, m_piecePixelSize);
+    QPalette palette = dummyProgressBar.palette();
+    palette.setColor(QPalette::Highlight, Qt::transparent);
+    dummyProgressBar.setValue(100);
+    dummyProgressBar.render(&pixmap);
+    m_pieceFinishedImage = pixmap.toImage();
+    dummyProgressBar.setPalette(palette);
+    dummyProgressBar.render(&pixmap);
+    m_pieceNeededImage = pixmap.toImage();
+}
+
+PiecesMapWidget::~PiecesMapWidget()
+{
+}
+
+QSize PiecesMapWidget::sizeHint() const
+{
+    QSize hint = QWidget::sizeHint();
+    if(m_torrent == nullptr) return hint;
```

Comment:
Invalid coding style.
And in many places below. You should consider qBittorrent Coding Guidelines before contributing.

Reply:
I read the guidelines today and made fixes. I'll trigger a code review at the end of the week.